### DISCUSSION TITLE
Output modal: Execute other shell commands

### DISCUSSION THREAD

**Discussion by FelipeRearden** [Taitava/obsidian-shellcommands#141] [Ideas]
Hello @Taitava !!!!!!!!!

Since the first day of testing I think about this new feature. I know it's a bit crazy idea, but let me share with you to see what you think :)

Would be amazing to be able to execute an SC from the modal output window. This way we could create more complex workflows starting from the modal and using the output of modal as input for the next SC command. 

### My Idea 

If we have a new event we could select the SC commands that will be available in the Modal window.  It's an event that looks like File Menu and Folder Menu.

### My Doubt 

The only thing that I don't know yet is how to display this new item in the Modal window. 

Since we could have more than one SC with the event toggle ON. I'm not sure yet how to display these list in the modal window in a way that we could have keyboard navigation as well :(

Have a great day!!!!

**Comment by Taitava** [comment]
Thank you for the idea! 🌞 

If I understood correctly, the idea is to:
1. Execute a shell command which outputs text to modal.
2. In the modal, there would be some element where a user could select another shell command to be executed.
3. The other shell command will receive the first command's output some way:
  - Via a variable?
  - Or via [`stdin` (not implemented yet)](https://github.com/Taitava/obsidian-shellcommands/discussions/89)?

Did I understood this correctly? 🙂 If yes, is it ok for you if I rename this discussion to: _Chainable shell commands_?

---

Do you have an example how you would use this in practise? My question is mainly this: Do we need to split shell commands into two separate shell commands? Would a basic piping work instead, e.g. `ls | grep abc` first gets a list of files and directories in the current directory, and then filters out all but those whose name contains `abc`. This was just a really small and simple example.

Also, I think one _competitor idea_ for this is this one of yours: https://github.com/Taitava/obsidian-shellcommands/discussions/127 . I know these are different kinds of solutions, but both of them can be applied for similar kinds of things: executing one shell command after another, while passing data between them. They don't necessarily rule out each other, but if both will be implemented, a priority order needs to be decided for the development. 🙂 

---

If something like this will be planned, we can continue thinking about what kind of user interface element will be used to select the next shell command. I don't have an idea for it yet, but we'll figure out something. But first we should come up with a use case for this.

**Reply by FelipeRearden** [reply]
Hello @Taitava !!!!

Before answer your questions, let me rephrase my idea:

1- Extend the set of features if the Modal Output adding a new "button" the lets the user run another SC from the Modal UI

2- The criteria to select which SC will be available in the modal UI would be by creating a new EVENT (the same way that we have File Menu and Folder Menu to filter which SC command will be available in the context menu)

3- I'm not sure yet how to display this new feature in the Modal UI :(

> If I understood correctly, the idea is to:
> 
> 1. Execute a shell command which outputs text to modal.


Yes or just run a SC that opens the Modal empty for me to insert text. 

 
> 2. In the modal, there would be some element where a user could select another shell command to be executed.

Yes !!!!!


> 3. The other shell command will receive the first command's output some way:

No !!!! That's the interesting point here ...

In my opinion, would be amazing to have an option to create a workflow with the Modal Output where I can execute another SC. 

Without a restriction to have to use the content in the modal (text) as an input for the next SC action. 

Below I will show two workflows for this new feature 
- One using the modal content 
- One without using the Modal content 

> Do you have an example how you would use this in practise? My question is mainly this: Do we need to split shell commands into two separate shell commands? Would a basic piping work instead, e.g. `ls | grep abc` first gets a list of files and directories in the current directory, and then filters out all but those whose name contains `abc`. This was just a really small and simple example.


### 1- Send text to an specific Note 

Step 1: Open an SC echo Journal [output=Modal]
Step 2: Write my angry journal 
Step 3: Copy the content of the modal to the clipboard using the `Clipboard` button in the bottom of the Modal UI
Step 4: Using the SC command echo {{clipboard}} > angry_journal.md to insert the content in the booth of a specific note.

### 2- Send Content to Another App Outside Obsidian

Step 1: Open an SC echo Feature Request [output=Modal]
Step 2: Write an E-mail
Step 3: Copy the content of the modal to the clipboard using the `Clipboard` button in the bottom of the Modal UI
Step 4: Using the SC command to open my git link to write create a new Issue on SC github. 
Step5: Paste the content in github

> Also, I think one _competitor idea_ for this is this one of yours: https://github.com/Taitava/obsidian-shellcommands/discussions/127 . I know these are different kinds of solutions, but both of them can be applied for similar kinds of things: executing one shell command after another, while passing data between them. They don't necessarily rule out each other, but if both will be implemented, a priority order needs to be decided for the development. 🙂 

100% agree @Taitava !!!!

> If something like this will be planned, we can continue thinking about what kind of user interface element will be used to select the next shell command. I don't have an idea for it yet, but we'll figure out something. But first we should come up with a use case for this.

Okay @Taitava !!!!! :) 

Let me know if you need anymore information!!!!!!

Have a great day!!!!!!

**Reply by Taitava** [reply]
> Before answer your questions, let me rephrase my idea:
> 
> 1- Extend the set of features if the Modal Output adding a new "button" the lets the user run another SC from the Modal UI

Yea, that's what I was thinking 🙂. 

> 
> 2- The criteria to select which SC will be available in the modal UI would be by creating a new EVENT (the same way that we have File Menu and Folder Menu to filter which SC command will be available in the context menu)

Yes, that might be one option for designing this. Another one might be to make it a bit more automatic: passing the output to the next shell command requires some kind of variable (at least until #89 gets implemented), so the list could consist of all shell commands that use the variable. The variable might be named `{{last_output}}` or something else - I haven't decided yet.

Or then make the variable name customisable by the user - then the list of allowed next shell commands can be filtered to only those that use the specific custom variable. So you could have two (or more) workflows that use output from previous shell commands, with different lists of shell commands that can be executed next. One workflow might use a variable named `{{custom_outputted_file_name}}`, and have only file management related shell commands in the list of next commands. Another might use `{{custom_outputted_date}}` and a list of shell commands that do something with dates.

Please let me know if anything of this was confusing? 🙂 (Below I will comment on the ability to execute shell commands that do not take the output of a previous shell command so directly.)

> Yes or just run a SC that opens the Modal empty for me to insert text.

Note that there will also be #17 that will allow collecting input from the user, and that will happen _before_ execution, so there's no need to have a shell command opening up an empty output modal.




> ### 1- Send text to an specific Note
> 
> Step 1: Open an SC echo Journal [output=Modal]
> Step 2: Write my angry journal 
> Step 3: Copy the content of the modal to the clipboard using the `Clipboard` button in the bottom of the Modal UI 
> Step 4: Using the SC command echo {{clipboard}} > angry_journal.md to insert the content in the booth of a specific note.

So, this would not use the modal's text field's content _dfirectly_. It would use it via the clipboard, so it would require you to do two clicks: first the clipboard button, and then click the shell command you want to execute next. It might work ok.

> 
> ### 2- Send Content to Another App Outside Obsidian
> 
> Step 1: Open an SC echo Feature Request [output=Modal] 
> Step 2: Write an E-mail 
> Step 3: Copy the content of the modal to the clipboard using the `Clipboard` button in the bottom of the Modal UI 
> Step 4: Using the SC command to open my git link to write create a new Issue on SC github. 
> Step5: Paste the content in github

Yea, here you would paste manually from the clipboard, so there's no need for direct access to the previous shell command's output. Thank you for these examples! 🙂 

---

I think this could be developed in steps:
1. phase: Develop it so that the list of next shell commands consists of shell commands that contain the custom variables mentioned above in this message.
2. phase: Develop a way to include other shell commands that do not use the variables.

**Comment by Taitava** [comment]
# I changed the discussion's title
Old title: _[Idea] New Event: Add to Modal Output_
New title: _Output modal: Execute other shell commands_

**Comment by FelipeRearden** [comment]
Hello @Taitava !!!!

Another workflow that I’m facing now related to use the content of the Ask for Execution Modal

Action = Duplicate a folder and the content inside (or a File) + Change the Name of the duplicated Folder (or File)
- Folder C:\Vault\Felipe
- Duplicate this folder to a new name C:\Vault\Jare

Workflow = I have a template folder (directory) that contains template notes for starting a new book notes workflow 

Every time that I’m starting a new book I have to duplicate this folder and change the name and number to the new book !

![35729EDB-08D7-45F5-93A8-CCE5E5DC80B4](https://user-images.githubusercontent.com/85180696/155854487-ca8203c2-2ce8-40b3-a20c-c57c50f73afa.jpeg)


SC 1 (Ask for Execution empty for me to type the name of the duplicated folder) 
`echo ` [output = AFE] [Event=Folder Menu + File Menu]

SC 2 (will be selected inside the AFE modal UI) (Windows 7 -cmd)
`XCOPY {{event_folder_path:absolute}}\* {{event_folder_path:absolute}}\{{modal_output}}.md \s \i` [output = ignore]

This way I could add this new commands to improve my File Management system with SC and Quick Explorer :)
- I have a long list of commands to File Management to do with our new features in 0.10 and 0.11 
- The main problem here is when I need to change the name of the thing that I’m managing (duplicate or move or new)

Have a great day!!!!

**Reply by Taitava** [reply]
Duplicating folders is nice idea! 🙂 

> SC 1 (Ask for Execution empty for me to type the name of the duplicated folder)
> `echo ` [output = AFE] [Event=Folder Menu + File Menu]

I still think having an empty echo shell command is not needed. You'll do just fine with the second shell command you introduced, when #17 / #37 is implemented. You can then attach a prompt to your shell command n:o 2 (and menu events) so that when you click the menu item to execute the shell command, the promp will open up and the shell command will be executed only after you have filled the prompt's fields and clicked an _Execute_ button.

You can also configure the prompt to show a custom message, e.g. _Insert a folder name_ or _Insert a file name_. And you can also decide the variable name where the answer will be stored, e.g. `{{_folder_name_given_by_user}}` (the variable name needs to start with `_` because it's a [custom variable](https://github.com/Taitava/obsidian-shellcommands/discussions/146).


Have a nice day, too! 🙂

**Reply by FelipeRearden** [reply]
Thanks @Taitava !!!!!

I'm glad that you liked ! 

Since I started creating workflows to file management, my idea was to share with you one more example where SC could be helpful and integrates with other plugin! 

Have a great day!!!!
